New resource `azurerm_virtual_machine_gallery_application_assignment`

myc2h6o opened the pull request:
Close #21094

myc2h6o commented:
Test result
![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/10579712/ba2c93fd-ab58-467b-914d-34b75e1409ac)

manicminer commented on internal/services/compute/parse/virtual_machine_gallery_application_assignment_id.go:
New resource IDs should be auto-generated by populating `resourceids.go` in the service package, and running `make generate`

myc2h6o replied:
this id is similar to [nat_gateway_public_ip_prefix_association](https://github.com/hashicorp/terraform-provider-azurerm/blob/fc0bfa18efa74a94c24b12eda48b00cb508c7d22/internal/services/network/parse/nat_gateway_public_ip_prefix_association.go) that combines two IDs together to make a new id, which I'm not able to use the resourceids.go to generate it. It looks like "id_1|id_2".

manicminer commented on internal/services/compute/parse/virtual_machine_gallery_application_assignment_id_test.go:
As above

myc2h6o replied:
similarly, I found the id and the test file hard to be auto generated, the test file is like [nat_gateway_public_ip_prefix_association_test](https://github.com/hashicorp/terraform-provider-azurerm/blob/fc0bfa18efa74a94c24b12eda48b00cb508c7d22/internal/services/network/parse/nat_gateway_public_ip_prefix_association_test.go)

manicminer commented on internal/services/compute/virtual_machine_gallery_application_assignment_resource.go:
```diff
@@ -0,0 +1,327 @@
+				return fmt.Errorf("checking for presence of existing %q: %+v", *virtualMachineID, err)
+			}
+
+			payload := virtualMachine.Model
```
Can we rename `payload` to something more readable like `virtualMachine`?

manicminer commented on internal/services/compute/virtual_machine_gallery_application_assignment_resource.go:
```diff
@@ -0,0 +1,327 @@
+			locks.ByID(virtualMachineID.ID())
+			defer locks.UnlockByID(virtualMachineID.ID())
+
+			virtualMachine, err := client.Get(ctx, *virtualMachineID, virtualmachines.GetOperationOptions{Expand: pointer.To(virtualmachines.InstanceViewTypesUserData)})
```
this could be `virtualMachineResult`

manicminer commented on internal/services/compute/virtual_machine_gallery_application_assignment_resource.go:
```diff
@@ -0,0 +1,327 @@
+				return fmt.Errorf("checking for presence of existing %q: %+v", id.VirtualMachineId, err)
+			}
+
+			payload := virtualMachine.Model
```
as above

manicminer commented on internal/services/compute/virtual_machine_gallery_application_assignment_resource.go:
```diff
@@ -0,0 +1,327 @@
+			locks.ByID(id.VirtualMachineId.ID())
+			defer locks.UnlockByID(id.VirtualMachineId.ID())
+
+			virtualMachine, err := client.Get(ctx, id.VirtualMachineId, virtualmachines.GetOperationOptions{Expand: pointer.To(virtualmachines.InstanceViewTypesUserData)})
```
as above

manicminer commented on internal/services/compute/virtual_machine_gallery_application_assignment_resource.go:
```diff
@@ -0,0 +1,327 @@
+				return fmt.Errorf("checking for presence of existing  %q: %+v", id, err)
+			}
+
+			payload := virtualMachine.Model
```
as above

manicminer commented on internal/services/compute/virtual_machine_gallery_application_assignment_resource.go:
```diff
@@ -0,0 +1,327 @@
+			locks.ByID(id.VirtualMachineId.ID())
+			defer locks.UnlockByID(id.VirtualMachineId.ID())
+
+			virtualMachine, err := client.Get(ctx, id.VirtualMachineId, virtualmachines.GetOperationOptions{})
```
as above

manicminer commented on website/docs/r/virtual_machine_gallery_application_assignment.html.markdown:
```diff
@@ -0,0 +1,123 @@
+
+Manages a Virtual Machine Gallery Application Assignment.
+
+~> **NOTE:** Gallery Application Assignment can be defined either directly on `azurerm_linux_virtual_machine` and `azurerm_windows_virtual_machine` resources, or using the `azurerm_virtual_machine_gallery_application_assignment` resource - but the two ways cannot be used together. If both are used against the same Virtual Machine, spurious changes will occur.
```
This note should also suggest the use of `ignore_changes` on the virtual machine resource for the `gallery_application` block.

```suggestion
~> **Note** Gallery Application Assignments can be defined either directly on `azurerm_linux_virtual_machine` and `azurerm_windows_virtual_machine` resources, or using the `azurerm_virtual_machine_gallery_application_assignment` resource - but the two approaches cannot be used together. If both are used with the same Virtual Machine, spurious changes will occur. It's recommended to use `ignore_changes` for the `gallery_application` block on the associated virtual machine resources, to avoid a persistent diff when using this resource.
```

manicminer commented on website/docs/r/linux_virtual_machine.html.markdown:
Similarly to the note on the new resource, we need to add a note to the `gallery_application` block here, warning about the conflict between the two, and recommending the use of `ignore_changes` when the new resource is used.

manicminer commented on website/docs/r/windows_virtual_machine.html.markdown:
Similarly to the note on the new resource, we need to add a note to the `gallery_application` block here, warning about the conflict between the two, and recommending the use of `ignore_changes` when the new resource is used.

manicminer commented on website/docs/r/virtual_machine_gallery_application_assignment.html.markdown:
```diff
@@ -0,0 +1,123 @@
+
+The following arguments are supported:
+
+* `gallery_application_version_id` - (Required) The ID of the Gallery Application Version. Changing this forces a new resourec to be created.
```
Spelling

```suggestion
* `gallery_application_version_id` - (Required) The ID of the Gallery Application Version. Changing this forces a new resource to be created.
```

manicminer commented on website/docs/r/virtual_machine_gallery_application_assignment.html.markdown:
```diff
@@ -0,0 +1,123 @@
+
+* `gallery_application_version_id` - (Required) The ID of the Gallery Application Version. Changing this forces a new resourec to be created.
+
+* `virtual_machine_id` - (Required) The ID of the Virtual Machine. Changing this forces a new resourec to be created.
```
```suggestion
* `virtual_machine_id` - (Required) The ID of the Virtual Machine. Changing this forces a new resource to be created.
```

manicminer reviewed (changes requested):
Thanks for this @myc2h6o. THis is mostly looking good, if you can look at the comments below, and if the tests pass, this should be good to merge.

myc2h6o commented:
@manicminer thanks for reviewing the pr! I've updated it according to your comments, please take a look. (the pr is also a little bit out sync with the go version so I rebased it to main as well)

myc2h6o commented on website/docs/r/windows_virtual_machine_scale_set.html.markdown:
```diff
@@ -358,7 +358,7 @@ A `gallery_application` block supports the following:
 * `configuration_blob_uri` - (Optional) Specifies the URI to an Azure Blob that will replace the default configuration for the package if provided. Changing this forces a new resource to be created.
 
-* `order` - (Optional) Specifies the order in which the packages have to be installed. Possible values are between `0` and `2,147,483,647`. Changing this forces a new resource to be created.
+* `order` - (Optional) Specifies the order in which the packages have to be installed. Possible values are between `0` and `2147483647`. Defaults to `0`. Changing this forces a new resource to be created.
```
This is the scale set resources, where these are marked as ForceNew. In vm they are not.

tombuildsstuff commented:
@WodansSon since (I believe) you're picking up @myc2h6o's PR - it appears there's a separate API for this [per this comment](https://github.com/hashicorp/terraform-provider-azurerm/issues/21094#issuecomment-1675246174) - rather than introducing a new Virtual Resource for this - could we instead introduce a new (regular) resource which uses that API instead?

myc2h6o commented:
@tombuildsstuff the user is currently using azapi_update to update the VM resource to add the gallery application property https://github.com/L-u-k-e-GIT/azure_vm_applications/blob/41d0f67427bf9039e21f7be2f3df6a59783d5f1d/vm_applications.tf#L12. The json which is put in that comment is the Azure deployment reference. There is no standalone API like VM extension currently for gallery application, so we may still need to go with this approach.

kgopi1 commented:
Hi, Any progress on this PR , still awaiting for this resource ..

katbyte reviewed (changes requested):
tests seem to have failed due to 
```
 virtual_machine_gallery_application_assignment_resource.go:14:2: cannot find module providing package github.com/hashicorp/go-azure-sdk/resource-manager/compute/2021-11-01/virtualmachines: import lookup disabled by -mod=vendor
[09:33:01][Step 4/5] 	(Go version in go.mod is at least 1.14 and vendor directory exists.)
```
?

myc2h6o commented:
The `compute` package has been upgraded so old packages could not be found any more. I have rebased to the latest main which resolved this issue, could you take another look @katbyte  Thanks!

> tests seem to have failed due to
> 
> ```
>  virtual_machine_gallery_application_assignment_resource.go:14:2: cannot find module providing package github.com/hashicorp/go-azure-sdk/resource-manager/compute/2021-11-01/virtualmachines: import lookup disabled by -mod=vendor
> [09:33:01][Step 4/5] 	(Go version in go.mod is at least 1.14 and vendor directory exists.)
> ```
> 
> ?

eclipse313 commented:
@katbyte Do you have time to take a quick look at it again?

katbyte reviewed (approved):
LGTM 🎱
